tools/elastic/migration/Migration.cs

```diff
@@ -0,0 +1,226 @@
+    public async Task RunUpAsync()
+    {
+        _builder.Logger.LogInformation("Applying migration {version}", this.Version);
+        await RunScriptsAsync(_builder, $"up{Path.DirectorySeparatorChar}pre");
+        await UpAsync(_builder);
+        await RunScriptsAsync(_builder, "up");
+        await RunScriptsAsync(_builder, $"up{Path.DirectorySeparatorChar}post");
+
+        var version = new MigrationVersion(this.Version, "");
+        var response = await _builder.Client.IndexAsync(version, id => id.Index(_builder.MigrationOptions.MigrationIndex));
+        if (!response.IsValid)
+        {
+            _builder.Logger.LogError(response.OriginalException, "Failed to add migration {version} to index.  Error: {error}", this.Version, response.ServerError);
+            throw response.OriginalException;
+        }
+    }
+
+    /// <summary>
+    /// Run the rollback migration to downgrade tot he prior version.
+    /// </summary>
+    /// <returns></returns>
+    public async Task RunDownAsync()
+    {
+        _builder.Logger.LogInformation("Rolling back migration {version}", this.Version);
+        await RunScriptsAsync(_builder, $"down{Path.DirectorySeparatorChar}pre");
+        await DownAsync(_builder);
+        await RunScriptsAsync(_builder, "down");
+        await RunScriptsAsync(_builder, $"down{Path.DirectorySeparatorChar}post");
+
+        var response = await _builder.Client.DeleteAsync<MigrationVersion>(this.Version, dd => dd.Index(_builder.MigrationOptions.MigrationIndex));
+        if (!response.IsValid)
+        {
+            var error = response.OriginalException ?? new ElasticsearchClientException("Failed to perform delete");
+            _builder.Logger.LogError(error, "Failed to remove migration {version} from index.  Error: {error}", this.Version, response.ServerError);
+            throw error;
+        }
+    }
+
+    /// <summary>
+    /// Run all the scripts found at the specified path for this version.
+    /// </summary>
+    /// <param name="builder"></param>
+    /// <param name="path"></param>
+    /// <returns></returns>
+    private async Task RunScriptsAsync(MigrationBuilder builder, string path)
+    {
+        var fullPath = $"{builder.MigrationOptions.MigrationsPath}{Path.DirectorySeparatorChar}{this.Version}{Path.DirectorySeparatorChar}{path}";
+        if (Directory.Exists(fullPath))
+        {
+            var files = Directory.GetFiles(fullPath).OrderBy(v => v).ToArray();
+
+            foreach (var file in files.Where(n => n.EndsWith(".json")))
+            {
+                // Perform each action in order.
+                builder.Logger.LogInformation("Applying migration step '{file}'", file);
+                using var stream = File.OpenRead(file);
+                var step = JsonSerializer.Deserialize<MigrationStep>(stream, builder.SerializerOptions) ?? throw new InvalidOperationException($"Failed to deserialize '{file}'");
+                if (step.Action == MigrationAction.CreateIndex)
```

Comment:
I'd love to come up with a better way to do this, but for now it does what we need.